homeassistant/helpers/llm.py

```diff
@@ -0,0 +1,240 @@
+        self,
+        name: str,
+        description: str,
+        parameters: vol.Schema = vol.Schema({}),
+        custom_serializer: Callable[[Any], Any] = default_custom_serializer,
+    ) -> None:
+        """Init the class."""
+        self.name = name
+        self.description = description
+        self.parameters = parameters
+        self.custom_serializer = custom_serializer
+
+    @cached_property
+    def specification(self) -> dict[str, Any]:
+        """Get the tool specification in OpenAPI-compatible format."""
+        return {
+            "name": self.name,
+            "description": self.description,
+            "parameters": convert(
+                self.parameters, custom_serializer=self.custom_serializer
+            ),
+        }
+
+    @abstractmethod
+    async def async_call(self, hass: HomeAssistant, tool_input: ToolInput) -> Any:
+        """Call the tool."""
+        raise NotImplementedError
+
+    def as_dict(self) -> dict[str, Any]:
+        """Get the tool specification in OpenAPI-compatible format."""
+        return self.specification
+
+    def __repr__(self) -> str:
+        """Represent a string of a Tool."""
+        return f"<{self.__class__.__name__} - {self.name}>"
+
+
+@callback
+def async_register_tool(hass: HomeAssistant, tool: Tool) -> None:
+    """Register an LLM tool with Home Assistant."""
+    if (tools := hass.data.get(DATA_KEY)) is None:
+        tools = hass.data[DATA_KEY] = {}
+
+    if tool.name in tools:
+        raise HomeAssistantError(f"Tool {tool.name} is already registered")
+
+    tools[tool.name] = tool
+
+
+@callback
+def async_remove_tool(hass: HomeAssistant, tool_name: str) -> None:
+    """Remove an LLM tool from Home Assistant."""
+    if (tools := hass.data.get(DATA_KEY)) is None:
+        return
+
+    tools.pop(tool_name, None)
+
+
+@callback
+def async_get_tools(hass: HomeAssistant) -> Iterable[Tool]:
+    """Return a list of registered LLM tools."""
+    if (tools := hass.data.get(DATA_KEY)) is None:
+        tools = {}
+
+    return tools.values()
+
+
+@callback
+async def async_call_tool(hass: HomeAssistant, tool_input: ToolInput) -> Any:
+    """Call a LLM tool, validate args and return the response."""
+    if (tools := hass.data.get(DATA_KEY)) is None:
+        tools = {}
+
+    tool = tools[tool_input.tool_name]
+
+    if tool_input.context is None:
+        tool_input.context = Context()
+
+    tool_input.tool_args = tool.parameters(tool_input.tool_args)
+
+    return await tool.async_call(hass, tool_input)
+
+
+class IntentTool(Tool):
+    """LLM Tool representing an Intent."""
+
+    def __init__(
+        self,
+        intent_type: str,
+        slot_schema: vol.Schema | None = None,
+    ) -> None:
+        """Init the class."""
+        if slot_schema is None:
+            slot_schema = vol.Schema({})
+        else:
+            slot_schema = vol.Schema(
+                {key: val.schema["value"] for key, val in slot_schema.schema.items()}
```

Comment:
Consider doing this outside of the tool (e.g. a helper function that converts an intent slot schema to a tool slot schema, then pases it to the intent tool constructor.  While subtle, the idea here is that the tool doesn't have to depend on details of the intent slot schema.  The constructor could also go away.

Reply:
Done. But we still need the constructor for the description.

Reply:
OK though it could also be passed as an argument.

Reply:
This section was also replaced with `effective_slot_schema`